kornia/augmentation/callbacks/base.py

```diff
@@ -0,0 +1,269 @@
+        self,
+        input: Tensor,
+        params: Dict[str, Tensor],
+        flags: Dict[str, Any],
+        inverse: Optional[Tensor] = None,
+    ):
+        """Called when `inverse_inputs` ends."""
+        ...
+
+    def on_inverse_masks_start(
+        self,
+        input: Tensor,
+        params: Dict[str, Tensor],
+        flags: Dict[str, Any],
+        inverse: Optional[Tensor] = None,
+    ):
+        """Called when `inverse_masks` begins."""
+        ...
+
+    def on_inverse_masks_end(
+        self,
+        input: Tensor,
+        params: Dict[str, Tensor],
+        flags: Dict[str, Any],
+        inverse: Optional[Tensor] = None,
+    ):
+        """Called when `inverse_masks` ends."""
+        ...
+
+    def on_inverse_classes_start(
+        self,
+        input: Tensor,
+        params: Dict[str, Tensor],
+        flags: Dict[str, Any],
+        transform: Optional[Tensor] = None,
+    ):
+        """Called when `inverse_classes` begins."""
+        ...
+
+    def on_inverse_classes_end(
+        self,
+        input: Tensor,
+        params: Dict[str, Tensor],
+        flags: Dict[str, Any],
+        transform: Optional[Tensor] = None,
+    ):
+        """Called when `inverse_classes` ends."""
+        ...
+
+    def on_inverse_boxes_start(
+        self,
+        input: Boxes,
+        params: Dict[str, Tensor],
+        flags: Dict[str, Any],
+        inverse: Optional[Tensor] = None,
+    ):
+        """Called when `inverse_boxes` begins."""
+        ...
+
+    def on_inverse_boxes_end(
+        self,
+        input: Boxes,
+        params: Dict[str, Tensor],
+        flags: Dict[str, Any],
+        inverse: Optional[Tensor] = None,
+    ):
+        """Called when `inverse_boxes` ends."""
+        ...
+
+    def on_inverse_keypoints_start(
+        self,
+        input: Keypoints,
+        params: Dict[str, Tensor],
+        flags: Dict[str, Any],
+        inverse: Optional[Tensor] = None,
+    ):
+        """Called when `inverse_keypoints` begins."""
+        ...
+
+    def on_inverse_keypoints_end(
+        self,
+        input: Keypoints,
+        params: Dict[str, Tensor],
+        flags: Dict[str, Any],
+        inverse: Optional[Tensor] = None,
+    ):
+        """Called when `inverse_keypoints` ends."""
+        ...
+
+    def on_forward_start(self, input: Tensor, params: Optional[Dict[str, Tensor]] = None, **kwargs: Any):
+        """Called when `forward` starts."""
+        ...
+
+    def on_forward_end(self, input: Tensor, params: Optional[Dict[str, Tensor]] = None, **kwargs: Any):
+        """Called when `forward` ends."""
+        ...
+
+    def on_inverse_start(self, input: Tensor, params: Optional[Dict[str, Tensor]] = None, **kwargs: Any):
+        """Called when `inverse` starts."""
+        ...
+
+    def on_inverse_end(self, input: Tensor, params: Optional[Dict[str, Tensor]] = None, **kwargs: Any):
+        """Called when `inverse` ends."""
+        ...
+
+    def on_sequential_forward_start(
```

Comment:
im a bit confused here since i see that this callback could be decoupled in two

1. AugmentationSequentialCallback (a good question is also if with callback we can change state and make nonlinear states)
   - handles the pipeline state like the forward  
2. AugmentationOpCallback
  - this specific to single augmentation.
  - but e.g you could setup the callbacks so that you can inject things in between the pipeline

I think also that the callback should be able to forward propagate results and have access to the main workflow state

Reply:
well, if I'm defining in the container, I would expect to have logs for the whole pipeline, and on the ops level, only for that op. But the public callback should be the same for both... otherwise, we'll overcomplicate the API usage

```python
AugmentationSequential(
    Normalize(..., callbacks=[WandbLogger])
    ColorJitter(..., callbacks=None),
    RandomAffine(..., callbacks=[WandbLogger]),
    RandomSolarize(..., callbacks=None),
    ...,
    callbacks=[WandbLogger]
)
```

I would like it to work within `ImageSequential` which I think is a pretty nice feature, but we don't stand out as much

Reply:
Sure, my point was more about that you might want a custom callback e.g to pipeline a variable from from normalize to random affine. The only dangerous thing I see now is that how to prevent that the callback in the Normalize implements something on the hooks belonging to the sequential commands. At internal level I think at least we should distinguish that

Reply:
I am wondering if we actually need an augmentation-wise level of logging, or we support it only for the sequential wrapper.

Reply:
yeap, i'm not sure when/why i would need that aside from debugging the aug